bayes_opt1/util.py

```diff
@@ -1,10 +1,277 @@
+        return np.asscalar(mu)
+    
+    dim = optimizer._space.bounds.shape[0] # size of state space (e.g. if we are working on R^3, n = 3) 
+    x_obs_temp = np.array([[res["params"]] for res in optimizer.res]) 
+    y_obs = np.array([res["target"] for res in optimizer.res])
+    # x_obs needs to be np.array of size (n,p)
+    # y_obs needs to be np.array of size (n,)
+    # see https://scikit-learn.org/stable/modules/generated/sklearn.gaussian_process.GaussianProcessRegressor.html
+    x_obs = x_obs_to_array(x_obs_temp, dim)
+
+    grid = grid_construction(optimizer._space.bounds, n_grid)
+    
+    temp = find_mean_max(x_obs,y_obs,gp,grid)
+    mean_nstar = temp[0]
+
+    grad_temp = []
+    mean, std= gp.predict(x.reshape(1,-1), return_std=True)
+    for j in range(J):
+        print("j = ",j+1)
+        with warnings.catch_warnings():
+            warnings.simplefilter("ignore")
+            temp = find_mean_max(x_obs,y_obs,gp,grid,mean,std,x)
+            mean_n1star = temp[0]
+            x_star_hat = grid[temp[1]].reshape(1,-1)
+            def func(w):
+                return mu_x_star_hat(x_obs,y_obs,gp,x_star_hat,mean,std,w)
+            
+            grad_temp.append(nd.Gradient(func)(x))
+    grad = np.zeros(dim)
+    for i in grad_temp:
+        grad = grad + i
+    grad = grad/J
+    return grad
+
+
 
+# Implementation algorithm 3 (Frazier): Knowledge gradient 
+def minimize_kg_sgd(R, T, a, pbounds, optimizer, gp):
 
-def acq_max(ac, gp, y_max, bounds, random_state, n_warmup=10000, n_iter=10):
+    KG = np.zeros(R)
+    init_lhs = our_lhs(R,pbounds)
+    dim = pbounds.shape[0]
+
+    if dim == 1:
+        
+        out = np.zeros(1)
+        x0 = np.zeros((T+1,R)) 
+        
+        for r in range(R):
+            # choose x0(r) uniformly random from A
+
+            x0[0,r] = init_lhs[r]
+
+            for t in range(1,T+1):
+                #print(x0[(t+1)*r].reshape(1,-1))
+                G = _kg4(x0[t-1,r], optimizer, gp, n_grid = 50, J = 5)
+                alpha = a/(a+t)
+                x0[t,r] = x0[t-1,r] + alpha*G
+            #print(type(x0[T-1,r]))        
+            KG[r] = _kg2(x0[T,r], optimizer, gp, n_grid = 50, J = 5)
+        
+        #print(type(x0[T, np.argmax(KG)]))
+        #print(np.array(x0[T, np.argmax(KG)]).shape)
+
+        out[0] = x0[T, np.argmax(KG)]
+        
+        return out 
+
+
+    else:
+        x0 = []  
+
+        for r in range(R):
+            # choose x0(r) uniformly random from A
+
+            x0.append(init_lhs[r].reshape(1,-1)) 
+            #print(init_lhs[r].reshape(1,-1))
+            for t in range(1,T+1):
+                print(x0[(t-1)+T*(r)].reshape(1,-1))
+                G = _kg4(x0[(t-1)+T*(r)].reshape(1,-1), optimizer, gp, n_grid = 50, J = 5)
+                alpha = a/(a+t)
+                x0.append(x0[(t-1)+T*(r)] + alpha*G)
+            #print(type(x0[T-1,r]))        
+            KG[r] = _kg2(x0[T*r].reshape(1,-1), optimizer, gp, n_grid = 50, J = 5)
```

Comment:
Ho parecchi dubbi su questo codice multi-dimensionale:
1) perché è separato dal caso uni-dimensionale? Ancora, sarebbe più gestibile se fosse un unico blocco
2) non capisco questa modalità di accesso agli elementi di `x0`, in cui appendete un array alla volta, ma a cui poi accedete con indici strani tipo `(t-1)+T*(r)`. Se avete bisogno, potete usare gli array tri-dimensionali: un asse per i restarts `r`, uno per le iterazioni `t`, e l'ultimo per le diverse coordinate dei punti. Si usano esattamente come quelli uni- o bi-dimensionali, per esempio:
```python 
x0 = np.zeros((3,4,2))  # dimensioni 3x4x2
print(x0)
print(x0[0,0,1])
```
Se proprio non volete usarli potete pensare a liste di array 2D (aka liste di matrici), che sono quasi la stessa cosa, ma sicuramente cambierei questo modo incomprensibile di tenere i valori in `x0`.

Direi quindi di gestire questo caso e inglobare, se si riesce, l'uni-dimensionale in questo, magari con un check simile a quello che avete aggiunto in `_kg4()`

Reply:
Abbiamo modificato la funzione... guardi l'ultimo commento!